### PR TITLE
Data inspector tool

### PR DESCRIPTION
**Pull request by KrystianKedron**
The first implementation of data inspector (task s4).
This PR implement the new approach to data inspector functionality similar like in the hightcharts framework. See demo: https://www.highcharts.com/demo/line-ajax.

### CODE REVIEW

File: taurus_pyqtgraph/datainspectortool.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        self._cb.setText('Data Inspector')
+        self._cb.toggled.connect(self._onTriggered)
+        self.setDefaultWidget(self._cb)
+
+        self.plot_item = None
+        self.Y2Axis = None
+        self.enable = False
+        self.data_inspector = None
+
+    def attachToPlotItem(self, plot_item, y2=None):
```
Comment (cpascual) [inline taurus_pyqtgraph/datainspectortool.py]: Unless I am missing something, I think that the y2 parameter is not used at all in this tool, so it should not be required (unless it is planned to be used later on, but I think it won't be necessary)

File: taurus_pyqtgraph/datainspectortool.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        :param plot_item: (PlotItem)
+        :param y2: (Y2ViewBox) instance of the Y2Viewbox attached to plot_item
+                   if the axis change controls are to be used
+        """
+        self.plot_item = plot_item
+        menu = plot_item.getViewBox().menu
+        menu.addAction(self)
+        self.Y2Axis = y2
+
+    def _onTriggered(self):
```
Comment (cpascual) [inline taurus_pyqtgraph/datainspectortool.py]: Very minor thing: maybe this method should be renamed to `_onToggled`?

File: taurus_pyqtgraph/datainspectormodel.py [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+
+        :param curve: is the reference to the current curve
+        :param point_index: index of picked point
+        :param i: the index of current curve
+        """
+
+        # Update itemas to make sure the older point is not still trigger
+        curve.updateItems()
+        scatterPoint = self.scatter[i].points()[point_index[0]]
+        scatterPoint.setSize(20)
```
Comment (cpascual) [inline taurus_pyqtgraph/datainspectormodel.py]: instead of hardcoding, make this `20` a class attribute (so that it can be changed)

File: taurus_pyqtgraph/datainspectormodel.py [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+
+                        parent = self.plot.curves[i]
+                        self.informationClouds[i].setParentItem(parent)
+
+                        self.scatter[i] = curve.scatter
+                        if curve.opts['symbol'] is None:
+                            # create the custom style of the point
+                            # if point style (symbol and size) not exist
+                            curve.setSymbol('o')
+                            curve.setSymbolSize(10)
```
Comment (cpascual) [inline taurus_pyqtgraph/datainspectormodel.py]: instead of hardcoding, make this `10` a class attribute (so that it can be changed)

File: taurus_pyqtgraph/datainspectormodel.py [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+                            self.custom_point_style_init = True
```
Comment (cpascual) [inline taurus_pyqtgraph/datainspectormodel.py]: this flag (`self.custom_point_style_init`) should be stored *per curve*. Otherwise, if a plot contains 2 curves, one having symbols and the other not having them, the symbols will be erased for **all** the curves when calling `reset_custom_style`

File: taurus_pyqtgraph/datainspectormodel.py [hunk 6]
```diff
@@ -0,0 +1,205 @@
+from datetime import datetime
+
+from pyqtgraph import TextItem, InfiniteLine
+import numpy as np
+from taurustrendset import TaurusTrendSet
+
+
+class DataInspectorModel(object):
```
Comment (cpascual) [inline taurus_pyqtgraph/datainspectormodel.py]: This comment is just an open question since it may be a matter of taste:

In its constructor, `DataInspectorModel` is implicitly "attaching" to the plot item (and effecting changes on the plot item). Likewise, it is implicitly calling `detach` in its destructor. 

I **personally** prefer to be explicit in situations like this. For example, I prefer to explicitly call `detach` instead of simply deleting the object in the `DataInspectorTool._onTriggered` method. The reasons are:
- follow Python Zen of ["explicit better than implicit"](https://www.python.org/dev/peps/pep-0020/)
- It gives the user of the class more control (e.g. the tool could instantiate the inspector just once and attach/detach it instead of instantiating/deleting it each time it is toggled)

File: taurus_pyqtgraph/datainspectormodel.py [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+    """
+    DataInspectorModel is a general class use to point-picking.
+    This class use a numpy library to finding the nearest point
+    from the :class:`pyqtgraph.InfiniteLine`.
+    Attach two widget to the plot:
+        - :class:`pyqtgraph.InfiniteLine` - to selecting point
+        - :class:`pyqtgraph.TextItem` - to showing the x,y value
+    Also this class blink the selected scatter point item.
+    """
+    def __init__(self, plot):
```
Comment (cpascual) [inline taurus_pyqtgraph/datainspectormodel.py]: This is a general comment: as a matter of coding style, we try to keep the **public** API to a minimum in order to facilitate refactorings later on. In parcatice this means that "by default" we make most members (attributes and methods) "protected" by sufixing them with `_` and only leave as "public" those which we explicitly want the users to access.
The philosophy is that it is easier to convert something protected into public if eventually needed than to have to go through a deprecation cycle and leave backwards-compatibility hacks if we want to refactor.

In practice this means that in this class members such as `self.scatter`, `self.informationClouds`, `self.custom_point_style_init`, `self.v_line`, `paintCloud()`, `triggerPoint()`, **etc.** would probably be better as "protected"

File: taurus_pyqtgraph/datainspectormodel.py [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+
+        if self.plot.sceneBoundingRect().contains(pos):
+            mousePoint = self.plot.vb.mapSceneToView(pos)
+
+            # Move InfiniteLine to the new position
+            self.v_line.setPos(mousePoint.x())
+
+            # Iteration over each curves in the plot and checking
+            # that mouse point if is near the x axis date
+            for i, curve in enumerate(filter(lambda o: not isinstance(o, TaurusTrendSet),
```
Comment (cpascual) [inline taurus_pyqtgraph/datainspectormodel.py]: why filtering out curves that are not trendsets?. IMHO this tool should work for any curve being displayed in the plot item.
Note that the idea is that the tools that we provide should be as generic (and taurus-independent) as possible.

File: taurus_pyqtgraph/datainspectormodel.py [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+        Removing the :class:`pyqtgraph.InfiniteLine` and
+        :class:`pyqtgraph.TextItem` from the plot
+        """
+        self.reset_custom_style()
+        self.plot.scene().removeItem(self.v_line)
+        for cloud in self.informationClouds.values():
+            self.plot.scene().removeItem(cloud)
+
+    @staticmethod
+    def getPointIndex(np_curve_data, mouse_point):
```
Comment (cpascual) [inline taurus_pyqtgraph/datainspectormodel.py]: This method should accept a parameter to set the tolerance (i.e. to be passed to `point_picker.checkPoint` as `roi`). The optimal value of that parameter should be calculated by `DataInspectorModel` taking into account the current scale. More specifically, IMHO the tolerance should be calculated considering the symbol size (e.g. 20 pixels) and translating it to a range in x axis coordinates)

File: taurus_pyqtgraph/datainspectormodel.py [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+        :param x: the new x position of the picked point
+        :param y: the new y position of the picked point
+        :param i: the index of current curve
+        """
+        self.informationClouds[i].setPos(x, y)
+        self.informationClouds[i].setHtml("<div style='%s'> "
+                                           "<span>x=%s "
+                                           "<span>y=%0.1f</span> "
+                                           "</div>" % (self.cloudStyle,
+                                                       self.timestampToTime(x),
```
Comment (cpascual) [inline taurus_pyqtgraph/datainspectormodel.py]: Converting to time should only be done if the x axis associated to the curve is a `DateAxisItem`

File: taurus_pyqtgraph/datainspectortool.py (hunk 1 above)
Comment (KrystianKedron) [reply taurus_pyqtgraph/datainspectortool.py]: Yes you are right. I correct this.

File: taurus_pyqtgraph/datainspectortool.py (hunk 2 above)
Comment (KrystianKedron) [reply taurus_pyqtgraph/datainspectortool.py]: I think a lot about name for this method -> onToggled is good idea.

File: taurus_pyqtgraph/datainspectormodel.py (hunk 3 above)
Comment (KrystianKedron) [reply taurus_pyqtgraph/datainspectormodel.py]: Should be provides the getter and setter for this variables?

File: taurus_pyqtgraph/datainspectormodel.py (hunk 5 above)
Comment (KrystianKedron) [reply taurus_pyqtgraph/datainspectormodel.py]: Ok, I update this.

File: taurus_pyqtgraph/datainspectormodel.py (hunk 6 above)
Comment (KrystianKedron) [reply taurus_pyqtgraph/datainspectormodel.py]: I update the code and now is more clear what happened. So this is the next reason. 
I like this approach. 
I hope this is what you mean.

File: taurus_pyqtgraph/datainspectormodel.py (hunk 7 above)
Comment (KrystianKedron) [reply taurus_pyqtgraph/datainspectormodel.py]: Ok, this is good point.
I think this is good place to ask this question, but I have problem:
How to properly call the method and variables?
Because Taurus use camelCase for named of the variables. 
Python PEP8 say that for this purpose should be use snake_case.
I use snace_case for named of the variables in the DataInspectorModel (I've noticed it now, but this is my habit) - should i change it to the camelCase? To keep consistency in the whole project? 
The same issue is with the named of the method.

File: taurus_pyqtgraph/datainspectormodel.py (hunk 8 above)
Comment (KrystianKedron) [reply taurus_pyqtgraph/datainspectormodel.py]: Now the sel.plot.curves don't contains only curves object. On the end of this collection is taurustrendset object.

I change this condition to be more agnostic and not creating the taurus dependency. Now I check that the object in this collection is the curve(plotdataitem object).

What you think?

File: taurus_pyqtgraph/datainspectormodel.py (hunk 10 above)
Comment (KrystianKedron) [reply taurus_pyqtgraph/datainspectormodel.py]: Ok, I implement this issue like you say, but now is dependecy to the custom AxisItem 
(DataAxisItem implemented by taurus pyqtgraph - so this isn't very agnostic?)

File: taurus_pyqtgraph/datainspectormodel.py (hunk 3 above)
Comment (cpascual) [reply taurus_pyqtgraph/datainspectormodel.py]: I would not bother for now. We can la add them later on if we need them. (the idea is to only create the minimum necessary public API, so that we do not need to maintain it afterwards)

File: taurus_pyqtgraph/datainspectormodel.py (hunk 7 above)
Comment (cpascual) [reply taurus_pyqtgraph/datainspectormodel.py]: the rule in taurus (which IMHO should be followed for taurus_pyqtgraph as well) is to make an exception to PEP8 case recommendation for consistency **in the Qt-related modules**. Any module that is not Qt-related, should follow the general PEP8 case rules. In the case of taurus_pyqtgraph, I think most of it will be Qt-related.

Note that taurus itself is pretty inconsistent since this rule was decided much later than the start of the project, and we decided not to refactor everything. But at least for new code we *try* to follow it

Comment (KrystianKedron) [reply taurus_pyqtgraph/datainspectormodel.py]: That explains a lot, 
Now I see the point.

File: taurus_pyqtgraph/datainspectormodel.py (hunk 9 above)
Comment (KrystianKedron) [reply taurus_pyqtgraph/datainspectormodel.py]: This is a good idea to optimize the algorytm finding the point on the curves.

I implemented very simple ROI calculation based on the pixel size and symbol size
(calculation relies on the minimum and maximum value in the X axis).

File: taurus_pyqtgraph/datainspectormodel.py [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+
+        if self.plot.sceneBoundingRect().contains(pos):
+            mousePoint = self.plot.vb.mapSceneToView(pos)
+
+            # Move InfiniteLine to the new position
+            self._v_line.setPos(mousePoint.x())
+
+            # Iteration over each curves in the plot and checking
+            # that mouse point if is near the x axis date
+            for i, curve in enumerate(filter(lambda o: isinstance(o, PlotDataItem),
```
Comment (cpascual) [inline taurus_pyqtgraph/datainspectormodel.py]: Is this filter really necessary? Maybe just do it for all items in self.plot.curves...

Comment (KrystianKedron) [reply taurus_pyqtgraph/datainspectormodel.py]: Yes, this feature is not necessary, 
but I only want to be sure that I iterate only over curves object, 
because like I said in this collection there are also other objects. 
I tested it without filtering and it worked fine, so I removed not necessary filtering, like you said.

File: taurus_pyqtgraph/datainspectormodel.py [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+    """
+    Class use to handle the data with x or y date from curve and checking
+    if the point is near or not (base on the roi) the current mouse point.
+    This class extend the :class:`numpy.ndarray` to handle the numpy array
+    with curve data.
+    """
+    def __new__(cls, *args, **kwargs):
+        return np.array(*args, **kwargs).view(PointPicker)
+
+    def checkPoint(self, point_to_check, roi):
```
Comment (cpascual) [inline taurus_pyqtgraph/datainspectormodel.py]: I think this is essentially the same idea already implemented in `numpy.isclose`

File: taurus_pyqtgraph/datainspectormodel.py [hunk 13]
```diff
@@ -0,0 +1,270 @@
...
+
+
+class PointPicker(np.ndarray):
+    """
+    Class use to handle the data with x or y date from curve and checking
+    if the point is near or not (base on the roi) the current mouse point.
+    This class extend the :class:`numpy.ndarray` to handle the numpy array
+    with curve data.
+    """
+    def __new__(cls, *args, **kwargs):
```
Comment (cpascual) [inline taurus_pyqtgraph/datainspectormodel.py]: Just a note on style. Maybe this is a justified use of `__new__`, but IMHO in most cases one can (should) use `__init__` instead for writing higher-level more pythonic code. See https://stackoverflow.com/a/674369

File: taurus_pyqtgraph/datainspectormodel.py [hunk 14]
```diff
@@ -0,0 +1,270 @@
...
+            for i, curve in enumerate(filter(lambda o: isinstance(o, PlotDataItem),
+                                             self.plot.curves)):
+
+                curve_data = curve.getData()
+                curve_x_data = curve_data[0]
+                curve_y_data = curve_data[1]
+
+                # Getting the nearest point in the x axis
+                # (if exist - empty array if not)
+                point_index = self._getPointIndex(curve_x_data, mousePoint.x())
```
Comment (cpascual) [inline taurus_pyqtgraph/datainspectormodel.py]: I think that this implementation could be simplified if we used the approach proposed [here](https://groups.google.com/forum/#!msg/pyqtgraph/HU4GsfzAxXU/H8UPYbPjAwAJ). 

I created [a very basic example](https://gist.github.com/cpascual/5620f913f83cae4be2a4f50057111e4c) which follows that approach and with which we would get rid of the `PointPicker`, `_calculateRoi` and `_getPointIndex`

Comment (KrystianKedron) [reply taurus_pyqtgraph/datainspectormodel.py]: This implementation is great!
I modify the datainspectormodel like in the example and did small changes:

- To prevent shaking the graph on an automatic scale in` _createLabels` I use `setParentItem` method on the `text_item`. In this case pyqtgraph doesn't take into account the text_item and not adjust the scale (this couse the shaking the graph when we pick the point near the border of the plot).

- I add the point trigger logic, and also if needed converting the x value to the time.
Also I fromat the y value to a defined number of decimal places (like in y_format variable).

**Comment by cpascual** [comment]
I did some refactoring and updated with the latest master.
IMHO this is ready to be integrated

**Comment by cpascual** [comment]
Idea for improvement: Instead of showing x,y in each label, the x value could be an extra tag at the base (or top) of the vertical line

**Comment by cpascual** [comment]
The current implementation seems enough as a first step, hence I am merging it. However it should be improved with (at least) y2-axis support
